app/controllers/concerns/saml_idp_auth_concern.rb

```diff
@@ -45,21 +45,29 @@ def check_sp_active
   end
 
   def validate_service_provider_and_authn_context
-    @saml_request_validator = SamlRequestValidator.new
+    return if result.success?
+
+    analytics.saml_auth(
+      **result.to_h.merge(request_signed: saml_request.signed?),
+    )
+    render 'saml_idp/auth/error', status: :bad_request
+  end
 
-    @result = @saml_request_validator.call(
+  def result
+    @result ||= @saml_request_validator.call(
       service_provider: saml_request_service_provider,
       authn_context: requested_authn_contexts,
       authn_context_comparison: saml_request.requested_authn_context_comparison,
       nameid_format: name_id_format,
     )
+  end
 
-    return if @result.success?
-
-    analytics.saml_auth(
-      **@result.to_h.merge(request_signed: saml_request.signed?),
-    )
-    render 'saml_idp/auth/error', status: :bad_request
+  def validate_and_create_saml_request_object
+    # this saml_idp method creates the saml_request object used for validations
+    validate_saml_request
```

Comment:
Somewhat related to my previous comment and probably out of scope for what we want to address here, but would we want to extract the other "validations" occurring in [`validate_saml_request`](https://github.com/18F/saml_idp/blob/272586f41eba7fe5074178bbcf206472f6ca82fc/lib/saml_idp/controller.rb#L21-L32) to happen within our validator class / "bad request" handling as well? Specifically the `valid_saml_request?` check and `Nokogiri::XML::SyntaxError` rescue in that method. 

Imagining something like...

```rb
def saml_request_validator
  saml_request = Request.from_deflated_request(raw_saml_request, get_params: params)
  if saml_request.valid?
    SamlRequestValidator.new # ...
  else
    SamlRequestValidator.new # ...
  end
rescue SamlIdp::XMLSecurity::SignedDocument::ValidationError
  SamlRequestValidator.new # ...
rescue Nokogiri::XML::SyntaxError
  SamlRequestValidator.new # ...
end
```

Raising here only in case it would affect how we'd want to approach this generally, such as the snippet above avoiding the `before_action` altogether.

Reply:
yes, long-term we definitely want to extract other validation errors, as well as specific certificate signature errors. (you can see some code i wrote to address that here: https://github.com/18F/saml_idp/pull/86). 

however, there are a lot of unknowns around saml assertion changes (and can really balloon quickly in complexity), and this ticket is fairly tightly scoped to address the 500s that have been occurring regularly, so i do think this suggestion is out of scope of this particular ticket. better errors/clearer code is something i am hoping to have added as part of the protocols team roadmap when it spins up!